### PR TITLE
New packet format & optimizations

### PR DESCRIPTION
**Pull request by chhwang**

### CODE REVIEW

File: include/mscclpp/semaphore_device.hpp [hunk 1]
```diff
@@ -20,14 +20,22 @@ struct Host2DeviceSemaphoreDeviceHandle {
   /// @return true if the host has signaled.
   MSCCLPP_DEVICE_INLINE bool poll() {
     bool signaled = (atomicLoad(inboundSemaphoreId, memoryOrderAcquire) > (*expectedInboundSemaphoreId));
-    if (signaled) (*expectedInboundSemaphoreId) += 1;
+    if (signaled) {
+      (*expectedInboundSemaphoreId) += 1;
+    } else {
+      // TODO: MSRCHA-371
+      atomicStore(&inboundSemaphoreId[1], uint64_t{0}, memoryOrderRelaxed);
+    }
```
Comment (Binyang2014) [inline include/mscclpp/semaphore_device.hpp:28]: Seems we can remove this and following workaround for MSRCHA-371

File: test/mscclpp-test/allgather_test.cu [hunk 2]
```diff
@@ -288,6 +290,215 @@ __global__ void allgather4(int rank, int worldSize, int nRanksPerNode, size_t ne
...
+  const size_t unitBytes = unitBytesPerWarp * nWarp;
+  const size_t nLoop = bytes / unitBytes;
+
+  if (nLoop > 0) {
+    // First loop unrolling
+    const size_t peerIdx = wid % nPeer;
+    const size_t remoteRankLocalIndex = (peerIdx < rank ? peerIdx : peerIdx + 1);
+    const size_t offset = bytesPerGPU * remoteRankLocalIndex + (wid / nPeer) * unitBytesPerWarp;
+    smChans[peerIdx].get<16, false>(offset, unitBytesPerWarp, lid, WARP_SIZE);
+  }
```
Comment (Binyang2014) [inline test/mscclpp-test/allgather_test.cu:331]: Why do we need first loop unrolling? Any performance gains?

File: test/mscclpp-test/allgather_test.cu [hunk 3]
```diff
@@ -288,6 +290,215 @@ __global__ void allgather4(int rank, int worldSize, int nRanksPerNode, size_t ne
...
+
+  // double buffering
+  const size_t scratchOffset = (flag & 1) ? 0 : bytesPerGPU * nRanksPerNode * 2;
+
+  if (nLoop > 0) {
+    // First loop unrolling
+    const size_t peerIdx = wid % nPeer;
+    const size_t offset = bytesPerGPU * rank + (wid / nPeer) * unitBytesPerWarp;
+    smChans[peerIdx].putPackets(scratchOffset + offset * 2, offset, unitBytesPerWarp, lid, WARP_SIZE, flag);
+  }
```
Comment (Binyang2014) [inline test/mscclpp-test/allgather_test.cu:446]: Also the reason for this unrolling

File: test/mscclpp-test/allgather_test.cu (hunk 2 above)
Comment (chhwang) [reply test/mscclpp-test/allgather_test.cu:331]: This improved small message latencies on AMD

File: include/mscclpp/semaphore_device.hpp (hunk 1 above)
Comment (chhwang) [reply include/mscclpp/semaphore_device.hpp]: Thanks, reverted this change.

**Review by Binyang2014 — APPROVED** [review]

**Pull request merged**